src/blockrelay/graphene_set.cpp

```diff
@@ -356,6 +354,134 @@ std::vector<uint64_t> CGrapheneSet::Reconcile(std::set<uint64_t> &receiverSet, c
+    double senderBloomFpr,
+    uint64_t nLowerBoundTruePositives)
+{
+    return (nSenderFilterPositiveItems - nLowerBoundTruePositives) /
+               ((nReceiverUniverseItems - nLowerBoundTruePositives) * senderBloomFpr) -
+           1;
+}
+
+double CGrapheneSet::TruePositiveProbability(uint64_t nSenderFilterPositiveItems,
+    uint64_t nReceiverUniverseItems,
+    double senderBloomFpr,
+    uint64_t nLowerBoundTruePositives)
+{
+    double margin = TruePositiveMargin(
+        nSenderFilterPositiveItems, nReceiverUniverseItems, senderBloomFpr, nLowerBoundTruePositives);
+    double margin_plus_1 = margin + 1;
+
+    return pow(exp(margin) / pow(margin_plus_1, margin_plus_1),
+        (nReceiverUniverseItems - nLowerBoundTruePositives) * senderBloomFpr);
+}
+
+uint64_t CGrapheneSet::LowerBoundTruePositives(uint64_t nTargetItems,
+    uint64_t nSenderFilterPositiveItems,
+    uint64_t nReceiverUniverseItems,
+    double senderBloomFpr,
+    double successRate)
+{
+    // x* in the graphene paper
+
+    if (nSenderFilterPositiveItems == 0)
+        return 0;
+
+    uint64_t nLowerBoundTruePositives = 0;
+    double prob = TruePositiveProbability(
+        nSenderFilterPositiveItems, nReceiverUniverseItems, senderBloomFpr, nLowerBoundTruePositives);
+
+    while (prob <= (1 - successRate) &&
+           ((int)nLowerBoundTruePositives <= (int)std::min(nSenderFilterPositiveItems, nTargetItems)))
+    {
+        nLowerBoundTruePositives += 1;
+        prob += TruePositiveProbability(
+            nSenderFilterPositiveItems, nReceiverUniverseItems, senderBloomFpr, nLowerBoundTruePositives);
+    }
+
+    return (uint64_t)std::max(0, (int)(nLowerBoundTruePositives - 1));
+}
+
+double CGrapheneSet::FalsePositiveMargin(uint64_t nLowerBoundTruePositives,
+    uint64_t nReceiverUniverseItems,
+    double senderBloomFpr,
+    double successRate)
+{
+    // delta in the graphene paper
+    double log_b = log(1 - successRate);
+    double s = -log_b / ((nReceiverUniverseItems - nLowerBoundTruePositives) * senderBloomFpr);
+
+    return 0.5 * (s + sqrt(pow(s, 2) + 8 * s));
+}
+
+uint64_t CGrapheneSet::UpperBoundFalsePositives(uint64_t nTargetItems,
+    uint64_t nSenderFilterPositiveItems,
+    uint64_t nReceiverUniverseItems,
+    double senderBloomFpr,
+    double successRate)
+{
+    // y* in the graphene paper
+    uint64_t nLowerBoundTruePositives = LowerBoundTruePositives(
+        nTargetItems, nSenderFilterPositiveItems, nReceiverUniverseItems, senderBloomFpr, successRate);
+    double margin = FalsePositiveMargin(nLowerBoundTruePositives, nReceiverUniverseItems, senderBloomFpr, successRate);
+
+    return (uint64_t)std::min((double)nSenderFilterPositiveItems,
+        (1 + margin) * (nReceiverUniverseItems - nLowerBoundTruePositives) * senderBloomFpr);
+}
+
+double CGrapheneSet::FailureRecoveryFpr(uint64_t nItems,
+    uint64_t nReceiverUniverseItems,
+    double senderBloomFpr,
+    double successRate)
+{
+    return 1;
```

Comment:
add a comment here as to why the false positive rate during failure recovery is always 1

Reply:
That method was cruft; sorry about that. Removed.